### PR TITLE
GC Settings Changes

### PR DESCRIPTION
**Pull request by bcbrockway**
* Adds support for changing GC settings using environment variables. Default location and no. of files changed to /data/log/gc.log and 4 respectively.
* Minor changes to README.md.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -60,8 +56,14 @@ RUN set -e \
 # Export HTTP & Transport
 EXPOSE 9200 9300
 
+# Copy configuration
+COPY config /usr/share/elasticsearch/config
```
Comment (primeroz) [inline Dockerfile]: why was this moved ? this way anything we copy in won't have its permissions changed by the "RUN Chown" above

File: config/jvm.options [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+# flags to configure Netty
+-Dio.netty.noUnsafe=true
+-Dio.netty.noKeySetOptimization=true
+-Dio.netty.recycler.maxCapacityPerThread=0
+
+# log4j 2
+-Dlog4j.shutdownHookEnabled=false
+-Dlog4j2.disable.jmx=true
+
+-Djava.io.tmpdir=${ES_TMPDIR}
```
Comment (nickmintel) [inline config/jvm.options:72]: Comes from:

```
export ES_TMPDIR=`mktemp -d -t elasticsearch.XXXXXXXX`
```

This might cause issues - it will default to `/tmp` unless TMPDIR is set.

File: run.sh [hunk 3]
```diff
@@ -114,6 +114,23 @@ fi
 # configuration overrides
 # CONF directory and files need to be writable by the user running the container
 
+## GC Log Settings
+if [[ ! -z ${ES_GCLOG_FILE_COUNT} ]]; then
```
Comment (Code0x58) [inline run.sh:118]: _pedantry:_ I think shellcheck would complain about this and suggest `[[ -n "${ES_GCLOG_FILE_COUNT}" ]]` (there's a [pycharm plugin](https://plugins.jetbrains.com/plugin/10195-shellcheck) and shellcheck can be installed using companion kube/`update-tool`)

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+.idea/
```
Comment (Code0x58) [inline .gitignore]: user-specific (rather than project specific) things can be ignored locally:
```sh
cat > "$HOME/.config/git/gitignore" <<EOF
.idea/
EOF
git config --global core.excludesfile "$HOME/.config/git/gitignore"
```

**Review by Code0x58 — APPROVED** [review]

File: .gitignore (hunk 4 above)
Comment (bcbrockway) [reply .gitignore]: Thanks, removed!

File: config/jvm.options (hunk 2 above)
Comment (primeroz) [reply config/jvm.options:72]: This is ok for now, sure elastic could end up writing lots in tmp but i don't think we ever saw that as an issue 

https://www.elastic.co/guide/en/elasticsearch/reference/master/es-tmpdir.html

File: Dockerfile (hunk 1 above)
Comment (bcbrockway) [reply Dockerfile]: Ah, that was just to speed up the build while I was debugging. I searched the file to see if `config` was referenced anywhere else but didn't think to check for references to the parent dirs :roll_eyes:.

File: run.sh (hunk 3 above)
Comment (primeroz) [reply run.sh:118]: in our case though ES_GCLOG_FILE_COUNT is always initialized since we have a default.

if "someone" though was going to set it to "" ( empty ) in a kubernetes yaml, would that sitll match with [[ -n ]] ? 

-z would match that

File: config/jvm.options (hunk 2 above)
Comment (bcbrockway) [reply config/jvm.options:72]: I just copied the default file that gets created from inside the running container but you're right, it's probably something we should think about. Perhaps we could leave the default as it is but define `ES_TMPDIR=/data/tmp` in the Dockerfile?

File: run.sh (hunk 3 above)
Comment (Code0x58) [reply run.sh:118]: `[[ -n ]]` will fail, whereas `[[ -n "" ]]` won't match. I think with bash to detect unset vs. empty you need to do `[ -n "${var-}" ] && [ "${var-x}" = x ]` (the first `${var-}` is just to work with `set -o nounset`). Don't get caught out [by `:-` vs. `-`](https://wiki.bash-hackers.org/syntax/pe#use_a_default_value) like I have been

Comment (primeroz) [reply run.sh:118]: Unless this is a real issue i would merge it as it is. All the other checks we do are "-z" can you in case create a separate case for this change of "bash variable checks" ?

Comment (Code0x58) [reply run.sh:118]: no complaints here, we don't have linting in the lifecycle of this
